### PR TITLE
fix: autolink on edit

### PR DESCRIPTION
**Pull request by allroundexperts**
### Details

This PR removes the `autolink` parsing rule when a message is being edited. 

### Fixed Issues

$ https://github.com/Expensify/App/issues/9090

### Tests

1. Open any chat
2. send safe.directory/link message
3. Edit the message to remove the link
4. Verify that the link is removed when the message is sent

- [x] Verify that no errors appear in the JS console

#### Contributor (PR Author) Checklist
- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [ ] I added steps for local testing in the `Tests` section
    - [ ] I added steps for Staging and/or Production testing in the `QA steps` section
    - [ ] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [ ] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] iOS / native
    - [x] Android / native
    - [x] iOS / Safari
    - [x] Android / Chrome
    - [x] MacOS / Chrome
    - [x] MacOS / Desktop
- [x] I verified there are no console errors (if there’s a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [ ] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [ ] I verified that comments were added to code that is not self explanatory
    - [ ] I verified that any new or modified comments were clear, correct English, and explained “why” the code was doing something instead of only explaining “what” the code was doing.
    - [ ] I verified any copy / text shown in the product was added in all `src/languages/*` files
    - [ ] I verified any copy / text that was added to the app is correct English and approved by marketing by tagging the marketing team on the original GH to get the correct copy.
    - [ ] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named “index.js”. All platform-specific files are named for the platform the code supports as outlined in the README.
    - [ ] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/STYLE.md#jsdocs)) were followed
- [ ] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [ ] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/PR_REVIEW_GUIDELINES.md)
- [ ] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [ ] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [ ] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [ ] If a new component is created I verified that:
    - [ ] A similar component doesn't exist in the codebase
    - [ ] All props are defined accurately and each prop has a `/** comment above it */`
    - [ ] Any functional components have the `displayName` property
    - [ ] The file is named correctly
    - [ ] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [ ] The only data being stored in the state is data necessary for rendering and nothing else
    - [ ] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [ ] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [ ] All JSX used for rendering exists in the render method
    - [ ] The component has the minimum amount of code necessary for its purpose and it is
- [ ] If a new CSS style is added I verified that:
    - [ ] A similar style doesn’t already exist
    - [ ] The style can’t be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [ ] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [ ] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.

<details>
<summary><h4>PR Reviewer Checklist</h4></summary>

- [ ] I verified the correct issue is linked in the `### Fixed Issues` section above
- [ ] I verified testing steps are clear and they cover the changes made in this PR
    - [ ] I verified the steps for local testing are in the `Tests` section
    - [ ] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [ ] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [ ] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [ ] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I verified tests pass on **all platforms** & I tested again on:
    - [ ] iOS / native
    - [ ] Android / native
    - [ ] iOS / Safari
    - [ ] Android / Chrome
    - [ ] MacOS / Chrome
    - [ ] MacOS / Desktop
- [ ] I verified there are no console errors (if there’s a console error not related to the PR, report it or open an issue for it to be fixed)
- [ ] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [ ] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [ ] I verified that comments were added to code that is not self explanatory
    - [ ] I verified that any new or modified comments were clear, correct English, and explained “why” the code was doing something instead of only explaining “what” the code was doing.
    - [ ] I verified any copy / text shown in the product was added in all `src/languages/*` files
    - [ ] I verified any copy / text that was added to the app is correct English and approved by marketing by tagging the marketing team on the original GH to get the correct copy.
    - [ ] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named “index.js”. All platform-specific files are named for the platform the code supports as outlined in the README.
    - [ ] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/STYLE.md#jsdocs)) were followed
- [ ] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [ ] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/PR_REVIEW_GUIDELINES.md)
- [ ] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [ ] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [ ] If a new component is created I verified that:
    - [ ] A similar component doesn't exist in the codebase
    - [ ] All props are defined accurately and each prop has a `/** comment above it */`
    - [ ] Any functional components have the `displayName` property
    - [ ] The file is named correctly
    - [ ] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [ ] The only data being stored in the state is data necessary for rendering and nothing else
    - [ ] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [ ] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [ ] All JSX used for rendering exists in the render method
    - [ ] The component has the minimum amount of code necessary for its purpose and it is broken down into smaller components in order to separate concerns and functions
- [ ] If a new CSS style is added I verified that:
    - [ ] A similar style doesn’t already exist
    - [ ] The style can’t be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [ ] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [ ] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.

</details>

### QA Steps

1. Open any chat
2. send safe.directory/link message
3. Edit the message to remove the link
4. Verify that the link is removed when the message is sent

- [ ] Verify that no errors appear in the JS console

#### Web

https://user-images.githubusercontent.com/30054992/173205257-93cad309-a4f4-4c51-aff6-2ac689748fdb.mov

#### Mobile Web

https://user-images.githubusercontent.com/30054992/173205238-4a7226aa-65ec-4d9a-bad9-b7525cdf04f9.mp4

#### Desktop

https://user-images.githubusercontent.com/30054992/173205222-f92da9b3-0dee-4491-a3e7-f157a30624f4.mov

#### iOS

https://user-images.githubusercontent.com/30054992/173205198-ccb308b3-a701-4a83-8f46-b4bd12b11223.mp4

#### Android

https://user-images.githubusercontent.com/30054992/173205184-08e6a358-8f4c-4463-8034-d8f905a7dceb.mov

### CODE REVIEW

**Review by luacmartins — COMMENTED** [review]
@allroundexperts I'm a bit confused since this solution is quite different from what you originally proposed. Could you please clarify why that's the case?

**Comment by allroundexperts** [comment]
> @allroundexperts I'm a bit confused since this solution is quite different from what you originally proposed. Could you please clarify why that's the case?

@luacmartins How is it different? At the time of my proposal, the `replace` function of the `ExpensiMark` class did not accept a second argument. At the time of creating the PR, it accepts a second argument called `filterRules`. I'm just making use of that.

**Comment by allroundexperts** [comment]
> > @allroundexperts I'm a bit confused since this solution is quite different from what you originally proposed. Could you please clarify why that's the case?
> 
> @luacmartins How is it different? At the time of my proposal, the `replace` function of the `ExpensiMark` class did not accept a second argument. At the time of creating the PR, it accepts a second argument called `filterRules`. I'm just making use of that.

@luacmartins This is the commit that was missing in my original proposal.
https://github.com/Expensify/expensify-common/commit/8ed32ff7bd4012df1f2c7a45d884ba2cd3c265d3

**Comment by luacmartins** [comment]
Ah I see. Thanks for clarifying! I'll probably be able to review this later this week. Meanwhile, @mananjadhav feel free to leave your comments.

**Comment by mananjadhav** [comment]
I'll be reviewing this by max tomorrow. Some system issues and I am just setting the workspace on another machine.

**Review by mananjadhav — COMMENTED** [review]
Change looks fine. I'll test this today/tomorrow and update here.

File: src/libs/actions/Report.js [hunk 1]
```diff
@@ -1242,7 +1242,7 @@ Onyx.connect({
  */
 function editReportComment(reportID, originalReportAction, textForNewComment) {
     const parser = new ExpensiMark();
-    const htmlForNewComment = parser.replace(textForNewComment);
+    const htmlForNewComment = parser.replace(textForNewComment, {filterRules: _.map(_.filter(parser.rules, (({name}) => !['autolink'].includes(name))), ({name}) => name)});
```
Comment (mananjadhav) [inline src/libs/actions/Report.js:1245]:
```suggestion
    const htmlForNewComment = parser.replace(textForNewComment, { filterRules: _.filter(_.pluck(parser.rules, 'name'), name => name !== 'autolink')});
```
Can we simplify the expression?

**Review by mananjadhav — CHANGES_REQUESTED** [review]
One change requested (restructure the expression). Rest of it tests well. Once you are done @allroundexperts ping me I'll test once again.

File: src/libs/actions/Report.js (hunk 1 above)
Comment (luacmartins) [reply src/libs/actions/Report.js:1245]: I agree with simplifying the expression! Thanks for the suggestion @mananjadhav!

Comment (allroundexperts) [reply src/libs/actions/Report.js]: Handled.

**Review by mananjadhav — DISMISSED** [review]
Thanks for the change @allroundexperts. LGTM.

@luacmartins 🎀 👀 🎀 C+ reviewed

**Comment by mananjadhav** [comment]
> Thanks for the change @allroundexperts. LGTM.
> 
> @luacmartins 🎀 👀 🎀 C+ reviewed

@luacmartins Bump ^

File: src/libs/actions/Report.js [hunk 2]
```diff
@@ -1242,7 +1242,7 @@ Onyx.connect({
  */
 function editReportComment(reportID, originalReportAction, textForNewComment) {
     const parser = new ExpensiMark();
-    const htmlForNewComment = parser.replace(textForNewComment);
+    const htmlForNewComment = parser.replace(textForNewComment, {filterRules: _.filter(_.pluck(parser.rules, 'name'), name => name !== 'autolink')});
```
Comment (luacmartins) [inline src/libs/actions/Report.js:1245]: Let's add a comment explaining why we are removing autolink as it's not obvious for someone not familiar with this issue.

**Review by luacmartins — COMMENTED** [review]
Sorry for the delay, I was ooo for the past few days. 

Changes test well. Left a small comment.

File: src/libs/actions/Report.js (hunk 2 above)
Comment (luacmartins) [reply src/libs/actions/Report.js:1245]: @allroundexperts bumping this!

Comment (allroundexperts) [reply src/libs/actions/Report.js:1248]: @luacmartins Updated.

Comment (luacmartins) [reply src/libs/actions/Report.js:1248]: Comment looks good, but linking is failing now

Comment (allroundexperts) [reply src/libs/actions/Report.js:1248]: @luacmartins Fixed.

**Review by luacmartins — APPROVED** [review]

**Comment by cead22** [comment]
Just for visibility, after merging this PR we found that editing a message to include a link made it so the link wasn't rendered. I'm not sure if that was considered in-scope for this PR, but mentioning it so in the future we always try to think about the flows for **creating and updating** messages when changing the code to render them

**Pull request merged**